lib/semantics/check-io.cc

```diff
@@ -96,15 +146,73 @@ void IoChecker::Enter(const parser::FileUnitNumber &spec) {
 void IoChecker::Enter(const parser::Format &spec) {
   SetSpecifier(IoSpecKind::Fmt);
   flags_.set(Flag::FmtOrNml);
-  if (std::get_if<parser::Star>(&spec.u)) {
-    flags_.set(Flag::StarFmt);
-  } else if (std::get_if<parser::Label>(&spec.u)) {
-    // Format statement format should be validated elsewhere.
-    flags_.set(Flag::LabelFmt);
-  } else {
-    flags_.set(Flag::CharFmt);
-    // TODO: validate compile-time constant format -- 12.6.2.2
-  }
+  std::visit(
+      common::visitors{
+          [&](const parser::Label &) { flags_.set(Flag::LabelFmt); },
+          [&](const parser::Star &) { flags_.set(Flag::StarFmt); },
+          [&](const parser::DefaultCharExpr &format) {
+            flags_.set(Flag::CharFmt);
+            const std::optional<std::string> constantFormat{
+                GetConstExpr<std::string>(format)};
+            if (!constantFormat) {
+              return;
+            }
+            // validate constant format -- 12.6.2.2
+            static constexpr int maxFormatErrors{3};
+            common::FormatError formatErrors[maxFormatErrors];
+            int errorCount{0};
+            switch (context_.GetDefaultKind(TypeCategory::Character)) {
```

Comment:
This looks a lot like the other place FormatValidator is called. Can the code be shared?

Reply:
It may be premature to consider doing this until more is known about handling multiple character kinds.  I've added a note about that in the TODO comments.